docs/bugs.json

```
@@ -1,4 +1,12 @@
 [
+    {
+        "name": "MemoryArrayCreationOverflow",
+        "summary": "The array length when creating a dynamic memory array is unrestricted and can overflow, potentially resulting in overlapping memory regions.",
+        "description": "No overflow checks were performed for the length of memory arrays during creation. In cases for which the memory size of an array in bytes rounded up to multiples of 32 is exactly 2^256-32, the memory allocation will overflow, potentially resulting in overlapping memory areas.",
```

```suggestion
        "description": "No runtime overflow checks were performed for the length of memory arrays during creation. In cases for which the memory size of an array in bytes (i.e. the length times 32) rounded up to multiples of 32 is exactly 2^256-32, the memory allocation will overflow, potentially resulting in overlapping memory areas.",
```
Isn't the condition `size is larger than 2**256-1`? Why exactly `2**256-32`?

---

We should also add: The length is still stored correctly, so this is usually not a problem since the newly created memory array will be iterated over resulting in out-of-gas for such large array sizes.

Hm, unless the length is overwritten because of the overlap :)

---

I changed the rest (I actually thought it needed to be exact, because we otherwise still zero-initialize correctly, yielding out-of-gas, but that wasn't true - that only saves the whole thing for ``bytes``) - let me check for a second how likely overwriting the length would be.

---

length*32 has to overflow, otherwise the zero-initialization will run out of gas...
We set the free memory pointer to ``add(oldFreeMem, add(mul(length, 32), 32)``... so since ``mul(length, 32)`` has already overflowed, this will still be > 32, so overwriting the length is unlikely. So I'll add that part to the description.

---

Updated.

---

I was thinking of overwriting the length by accessing the array at a very large index and thus wrapping the memory area, but this is rather unlikely to be possible in existing code.